### PR TITLE
Fix Windows IO Controller, do proper error checking

### PR DESCRIPTION
**Pull request by JosephusPaye**
This fixes the Windows IO Controller, addressing all TODOs and filling in unfinished code, with proper error checking.

All the tests now pass on Windows, including the ones that were failing in NUClearNet.js, which is what revealed the IO controller issues in the first place.

### CODE REVIEW

File: src/extension/IOController_Windows.hpp [hunk 1]
```diff
@@ -54,75 +62,113 @@ namespace extension {
...
+                }
             });
 
             on<Always>().then("IO Controller", [this] {
                 if (!shutdown) {
-
                     // Wait for events
-                    auto event = WSAWaitForMultipleEvents(
-                        static_cast<DWORD>(fds.size()), fds.data(), false, WSA_INFINITE, false);
+                    auto eventIndex = WSAWaitForMultipleEvents(
```
Comment (TrentHouliston) [inline src/extension/IOController_Windows.hpp:135]:
```suggestion
                    auto event_index = WSAWaitForMultipleEvents(
```

File: src/extension/IOController_Windows.hpp [hunk 2]
```diff
@@ -139,22 +185,37 @@ namespace extension {
                     }
                 }
 
-                // If not shudown
+                if (reactions_list_dirty || events_to_close.size() > 0) {
```
Comment (TrentHouliston) [inline src/extension/IOController_Windows.hpp:188]:
```suggestion
                if (reactions_list_dirty || !events_to_close.empty() > 0) {
```

File: src/extension/IOController_Windows.hpp [hunk 3]
```diff
@@ -139,22 +185,37 @@ namespace extension {
...
+                        for (auto& event : events_to_close) {
+                            if (!WSACloseEvent(event)) {
+                                throw std::system_error(
+                                    WSAGetLastError(), std::system_category(), "WSACloseEvent() failed");
+                            }
+                        }
 
-                // If dirty then update our list
+                        // Clear the queue of events to close
+                        events_to_close.resize(0);
```
Comment (TrentHouliston) [inline src/extension/IOController_Windows.hpp:202]: When you don't intend to reuse the memory use clear.
Clear will remove the backing memory, resize(0) leaves it untouched
```suggestion
                        events_to_close.clear();
```

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

File: src/extension/IOController_Windows.hpp [hunk 4]
```diff
@@ -34,18 +34,26 @@ namespace extension {
             // Startup WSA for IO
             WORD version = MAKEWORD(2, 2);
             WSADATA wsa_data;
-            WSAStartup(version, &wsa_data);
+
+            int startup_status = WSAStartup(version, &wsa_data);
+            if (startup_status != 0) {
+                throw std::system_error(startup_status, std::system_category(), "WSAStartup() failed");
```
[the remark this answers is not in the document]
Comment (JosephusPaye) [reply src/extension/IOController_Windows.hpp:40]: I'm not sure as well, and chose system to match the errors in IOController_Posix

File: src/extension/IOController_Windows.hpp (hunk 3 above)
Comment (JosephusPaye) [reply src/extension/IOController_Windows.hpp:202]: Nice, TIL!

File: src/extension/IOController_Windows.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply src/extension/IOController_Windows.hpp:40]: I checked, this is fine, it uses the same error category as the WSAGetLastError()

**Review by TrentHouliston — APPROVED** [review]

**Pull request merged**